File: ts/project/.sdk/tm/php/utility/Fetcher.php
<?php
declare(strict_types=1);

// ProjectName SDK utility: fetcher

class ProjectNameFetcher
{
    // Default User-Agent — many CDNs (notably Cloudflare) reject requests
    // with PHP's default UA (which file_get_contents doesn't even set),
    // returning 403 before the request reaches the origin. Set a Mozilla-
    // shaped UA so the SDK behaves like every other HTTP client by default.
    // Users can override by passing a User-Agent header in fetchdef.
    public const DEFAULT_USER_AGENT = 'Mozilla/5.0 (compatible; ProjectNameSDK/1.0)';

    public static function defaultHttpFetch(string $fullurl, array $fetchdef): array
    {
        $method_str = strtoupper($fetchdef['method'] ?? 'GET');
        $body_str = $fetchdef['body'] ?? null;
        $headers = $fetchdef['headers'] ?? [];

        $header_lines = [];
        $has_ua = false;
        foreach ($headers as $k => $v) {
            if (is_string($v)) {
                if (strcasecmp($k, 'user-agent') === 0) {
                    $has_ua = true;
                }
                $header_lines[] = "{$k}: {$v}";
            }
        }
        if (!$has_ua) {
            $header_lines[] = 'User-Agent: ' . self::DEFAULT_USER_AGENT;
        }

        // Prefer cURL when available — its header capture is reliable across
        // PHP versions, while file_get_contents + custom stream wrappers
        // don't propagate $http_response_header for user-defined wrappers
        // in PHP 8.3+.
        if (function_exists('curl_init')) {
            return self::curlFetch($fullurl, $method_str, $body_str, $header_lines);
        }

        $opts = [
            'http' => [
                'method' => $method_str,
                'ignore_errors' => true,
            ],
        ];

        if (is_string($body_str)) {
            $opts['http']['content'] = $body_str;
        }

        if (!empty($header_lines)) {
            $opts['http']['header'] = implode("\r\n", $header_lines);
        }

        $context = stream_context_create($opts);
        $response_body = @file_get_contents($fullurl, false, $context);

        $status = 0;
        $status_text = '';
        $resp_headers = [];

        $response_headers = function_exists('http_get_last_response_headers')
            ? http_get_last_response_headers()
            : ($http_response_header ?? []);

        if (is_array($response_headers)) {
            foreach ($response_headers as $header) {
                if (preg_match('/^HTTP\/\S+\s+(\d+)\s+(.*)$/i', $header, $matches)) {
                    $status = (int)$matches[1];
                    $status_text = trim($matches[2]);
                } else {
                    $parts = explode(':', $header, 2);
                    if (count($parts) === 2) {
                        $resp_headers[strtolower(trim($parts[0]))] = trim($parts[1]);
                    }
                }
            }
        }

        $json_body = null;
        if ($response_body !== false && $response_body !== '') {
            $decoded = json_decode($response_body, true);
            if (json_last_error() === JSON_ERROR_NONE) {
                $json_body = $decoded;
            }
        }

        return [
            [
                'status' => $status,
                'statusText' => $status_text,
                'headers' => $resp_headers,
                'json' => function () use ($json_body) { return $json_body; },
                'body' => $response_body !== false ? $response_body : '',
            ],
            null,
        ];
    }

    private static function curlFetch(string $fullurl, string $method, $body_str, array $header_lines): array
    {
        // GET-only guard for live tests — set by pub-live-test. We respect
        // it here at the SDK level since cURL bypasses PHP stream wrappers
        // (so the wrapper-based guard wouldn't see this call).
        if (getenv('VOXGIG_GETONLY_GUARD') === 'TRUE' && $method !== 'GET') {
            fwrite(STDERR, "[GET-ONLY GUARD] blocked {$method} {$fullurl}\n");
            return [
                [
                    'status' => 0,
                    'statusText' => 'Blocked',
                    'headers' => [],
                    'json' => function () { return null; },
                    'body' => '',
                ],
                "GET-only guard blocked {$method} {$fullurl}",
            ];
        }

        $ch = curl_init($fullurl);
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($ch, CURLOPT_HEADER, false);
        curl_setopt($ch, CURLOPT_FOLLOWLOCATION, false);
        curl_setopt($ch, CURLOPT_TIMEOUT, 30);
        curl_setopt($ch, CURLOPT_CONNECTTIMEOUT, 10);
        curl_setopt($ch, CURLOPT_CUSTOMREQUEST, $method);
        if (!empty($header_lines)) {
            curl_setopt($ch, CURLOPT_HTTPHEADER, $header_lines);
        }
        if (is_string($body_str)) {
            curl_setopt($ch, CURLOPT_POSTFIELDS, $body_str);
        }

        $resp_headers = [];
        curl_setopt($ch, CURLOPT_HEADERFUNCTION,
            function ($_ch, $hdr) use (&$resp_headers) {
                $trimmed = trim($hdr);
                if ($trimmed !== '') {
                    $resp_headers[] = $trimmed;
                }
                return strlen($hdr);
            });

        $response_body = curl_exec($ch);
        if ($response_body === false) {
            $err = curl_error($ch) ?: 'curl_exec failed';
            curl_close($ch);
            return [
                [
                    'status' => 0,
                    'statusText' => 'Error',
                    'headers' => [],
                    'json' => function () { return null; },
                    'body' => '',
                ],
                $err,
            ];
        }
        curl_close($ch);

        $status = 0;
        $status_text = '';
        $resp_kv = [];
        foreach ($resp_headers as $h) {
            if (preg_match('/^HTTP\/\S+\s+(\d+)\s*(.*)$/i', $h, $m)) {
                $status = (int)$m[1];
                $status_text = trim($m[2]);
            } else {
                $parts = explode(':', $h, 2);
                if (count($parts) === 2) {
                    $resp_kv[strtolower(trim($parts[0]))] = trim($parts[1]);
                }
            }
        }

        $json_body = null;
        if ($response_body !== '' && $response_body !== false) {
            $decoded = json_decode($response_body, true);
            if (json_last_error() === JSON_ERROR_NONE) {
                $json_body = $decoded;
            }
        }

        return [
            [
                'status' => $status,
                'statusText' => $status_text,
                'headers' => $resp_kv,
                'json' => function () use ($json_body) { return $json_body; },
                'body' => (string)$response_body,
            ],
            null,
        ];
    }

    public static function call(ProjectNameContext $ctx, string $fullurl, array $fetchdef): array
    {
        if ($ctx->client->mode !== 'live') {
            return [null, $ctx->make_error('fetch_mode_block',
                "Request blocked by mode: \"{$ctx->client->mode}\" (URL was: \"{$fullurl}\")")];
        }

        $options = $ctx->client->options_map();
        if (\Voxgig\Struct\Struct::getpath($options, 'feature.test.active') === true) {
            return [null, $ctx->make_error('fetch_test_block',
                "Request blocked as test feature is active (URL was: \"{$fullurl}\")")];
        }

        $sys_fetch = \Voxgig\Struct\Struct::getpath($options, 'system.fetch');

        // Treat null OR an empty stdClass/array as "no fetcher provided" and
        // fall through to the default HTTP fetcher. The options builder
        // sometimes materializes `system.fetch` as an empty stdClass even
        // when the user didn't set one — that's a placeholder, not a value.
        $is_empty_obj = ($sys_fetch instanceof \stdClass) && empty(get_object_vars($sys_fetch));
        $is_empty_arr = is_array($sys_fetch) && empty($sys_fetch);
        if ($sys_fetch === null || $is_empty_obj || $is_empty_arr) {
            return self::defaultHttpFetch($fullurl, $fetchdef);
        }
        if (is_callable($sys_fetch)) {
            return $sys_fetch($fullurl, $fetchdef);
        }

        return [null, $ctx->make_error('fetch_invalid', 'system.fetch is not a valid function')];
    }
}

File: ts/project/.sdk/tm/py/pkg/utility/fetcher.py
# ProjectName SDK utility: fetcher

from __future__ import annotations
import json
import threading
from http.cookiejar import DefaultCookiePolicy
from projectname_sdk.utility.voxgig_struct import voxgig_struct as vs


# Default User-Agent — many CDNs (notably Cloudflare) reject requests with
# a library's default UA, returning 403 before the request even reaches
# the origin. Set a Mozilla-shaped UA so the SDK behaves like every other
# HTTP client by default. Users can still override by passing a User-Agent
# header in fetchdef.
_DEFAULT_USER_AGENT = "Mozilla/5.0 (compatible; ProjectNameSDK/1.0)"


_SESSION = None
_SESSION_LOCK = threading.Lock()


def _session():
    # Pool connections without retaining response cookies between calls.
    global _SESSION
    if _SESSION is None:
        import requests
        with _SESSION_LOCK:
            if _SESSION is None:
                session = requests.Session()
                session.cookies.set_policy(DefaultCookiePolicy(allowed_domains=[]))
                _SESSION = session
    return _SESSION


def _default_http_fetch(fullurl, fetchdef):
    method = fetchdef.get("method", "GET")
    body_str = fetchdef.get("body")
    headers = fetchdef.get("headers", {})

    if not isinstance(body_str, str):
        body_str = None

    data = body_str.encode("utf-8") if body_str is not None else None

    req_headers = {}
    has_ua = False
    for k, v in headers.items():
        if k.lower() == "user-agent":
            has_ua = True
        req_headers[k] = v
    if not has_ua:
        req_headers["User-Agent"] = _DEFAULT_USER_AGENT

    # Manual redirects: fetchdef["redirect"] == "manual" surfaces a 3xx as
    # an ordinary response instead of auto-following it, which would replay
    # the request, headers included, against whatever host Location names.
    # Set by the station feature's transport middleware under an egress
    # hosts policy (mirrors the ts fetch option).
    kwargs = {"allow_redirects": fetchdef.get("redirect") != "manual"}

    proxy = fetchdef.get("proxy")
    if isinstance(proxy, str) and proxy:
        kwargs["proxies"] = {"http": proxy, "https": proxy}

    timeout = fetchdef.get("timeout")
    if isinstance(timeout, (int, float)) and timeout > 0:
        kwargs["timeout"] = timeout

    try:
        resp = _session().request(method, fullurl, data=data, headers=req_headers, **kwargs)
        body = resp.content.decode("utf-8")
    except Exception as e:
        return None, str(e)

    resp_headers = {}
    for k, v in resp.headers.items():
        resp_headers[k.lower()] = v

    json_body = None
    if len(body) > 0:
        try:
            json_body = json.loads(body)
        except Exception:
            pass

    status = resp.status_code
    status_text = resp.reason or ("OK" if status < 400 else "Error")

    return {
        "status": status,
        "statusText": status_text,
        "headers": resp_headers,
        "json": lambda: json_body,
        "body": body,
    }, None


def fetcher_util(ctx, fullurl, fetchdef):
    if ctx.client.mode != "live":
        return None, ctx.make_error("fetch_mode_block",
            'Request blocked by mode: "' + ctx.client.mode +
            '" (URL was: "' + fullurl + '")')

    options = ctx.client.options_map()
    if vs.getpath(options, "feature.test.active") is True:
        return None, ctx.make_error("fetch_test_block",
            'Request blocked as test feature is active'
            ' (URL was: "' + fullurl + '")')

    sys_fetch = vs.getpath(options, "system.fetch")

    if sys_fetch is None:
        return _default_http_fetch(fullurl, fetchdef)

    if callable(sys_fetch):
        return sys_fetch(fullurl, fetchdef)

    return None, ctx.make_error("fetch_invalid", "system.fetch is not a valid function")
